### ISSUE TITLE
TraceApi#createChildSpan: [no-plugin-name] Creating phantom child span [xxx] because the trace with root span [/some-route] has reached a limit of 1000 spans. This is likely a memory leak.

### ISSUE THREAD

**Issue by mrdulin** [googleapis/cloud-trace-nodejs#1044]
> @google-cloud/trace-agent ERROR TraceApi#createChildSpan: [no-plugin-name] Creating phantom child span [xxx] because the trace with root span [/some-route] has reached a limit of 1000 spans. This is likely a memory leak.

How can I solve this?

**Comment by kjin** [comment]
The number of child spans per root span should be similar to the number of outgoing RPCs per incoming request, so in general having 1000 child spans attached to a root span suggests that some child spans were attached to the wrong root spans.

I see the tag `[no-plugin-name]` which suggests that you are using the custom span API. Could you provide more details about how you're using it? In particular where the calls to `createChildSpan` are being made.

**Comment by mrdulin** [comment]
@kjin Thanks for reply. It's a little bit complex.

 I am using it with `GraphQL`, try to trace each field resolve process of SDL. I give you a screenshot for demonstration. For some privacy reason, I can't give you more business fields. But it's enough to explain my use case.

The trace span contain `GraphQL` `fields`  resolvers, `pg-query` and other api calls.
Another reason is I don't optimize the `GraphQL` N + 1 query issue, so there are many duplicated fields resolve. This increase the trace span count.

So, for above reasons. This error thrown.

![image](https://user-images.githubusercontent.com/17866683/59004357-f1132800-884b-11e9-9c4d-7abda61a7ad8.png)

**Comment by kjin** [comment]
Apologies for the delay -- I just got back from vacation.

The Stackdriver Trace API only supports 1000 spans per trace, which is why we set the limit to this number. To get rid of the warning you _could_ set `{ spansPerTraceHardLimit: Infinity }`, but this would only serve to send more data to the Stackdriver Trace service, and anything past the first 1000 spans will be lost. (This library doesn't have control over this 1000-span limit)

The `no-plugin-name` is a red herring, making it a plugin shouldn't have any effect on the error.

If it's possible to optimize the N+1 query issue I would recommend that first, as right now I can't think of any way of supporting GraphQL outside of decreasing the span count, unfortunately.
